wave_optimization/include/wave/optimization/factor_graph/factor_graph.hpp

```diff
@@ -0,0 +1,82 @@
+#ifndef WAVE_OPTIMIZATION_FACTOR_GRAPH_FACTOR_GRAPH_HPP
+#define WAVE_OPTIMIZATION_FACTOR_GRAPH_FACTOR_GRAPH_HPP
+
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <map>
+#include <set>
+#include <memory>
+
+#include "wave/utils/utils.hpp"
+#include "wave/optimization/factor_graph/factor.hpp"
+#include "wave/optimization/factor_graph/variable.hpp"
+
+namespace wave {
+
+class FactorGraph {
+ public:
+    std::multimap<std::shared_ptr<Variable>, std::shared_ptr<Factor>> graph;
+    std::set<std::shared_ptr<Variable>> variables;
+    std::vector<std::shared_ptr<Factor>> factors;
+
+    FactorGraph() : graph{}, variables{}, factors{} {}
+
+    template <typename V, typename T>
+    int addUnaryFactor(size_t at, const T &z) {
+        // store variable and factor
+        auto variable = std::make_shared<V>(at);
```

Comment:
`at` is an id of a factor graph variable, the choice of `at` can be anything, the most obvious choice is an integer number which we can increment easily by adding 1. Think of it as an id for every pose or landmark we are optimising.

The reason why its `size_t` is because it has a high bound, an `int` is represented by 4 bytes, and has a range of –2,147,483,648 to 2,147,483,647, where as `size_t` can store the maximum size of a theoretically possible object of any type.